libs/shared/flagd-core/src/lib/targeting/targeting.spec.ts

```diff
@@ -0,0 +1,161 @@
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should support lte operator', () => {
+    const input = {"sem_ver": ['v1.2.3', "<=", "1.2.3"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should support gte operator', () => {
+    const input = {"sem_ver": ['v1.2.3', ">=", "1.2.3"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should support gt operator', () => {
+    const input = {"sem_ver": ['v1.2.4', ">", "1.2.3"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should support major comparison operator', () => {
+    const input = {"sem_ver": ["v1.2.3", "^", "v1.0.0"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should support minor comparison operator', () => {
+    const input = {"sem_ver": ["v5.0.3", "~", "v5.0.8"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeTruthy()
+  });
+
+  it('should handle unknown operator', () => {
+    const input = {"sem_ver": ["v1.0.0", "-", "v1.0.0"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeFalsy()
+  });
+
+  it('should handle invalid inputs', () => {
+    const input = {"sem_ver": ["myVersion_1", "=", "myVersion_1"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeFalsy()
+  });
+
+  it('should validate inputs', () => {
+    const input = {"sem_ver": ["myVersion_2", "+", "myVersion_1", "myVersion_1"]}
+    expect(targeting.applyTargeting("flag", input, {})).toBeFalsy()
+  });
+})
+
+describe("fractional operator", () => {
+  let targeting: Targeting;
+
+  beforeAll(() => {
+    targeting = new Targeting();
+  })
+
+  it("should evaluate valid rule", () => {
+    const input = {
+      fractional: [
+        {"var": "key"},
+        ["red", 50],
+        ["blue", 50]
+      ]
+    }
+
+    expect(targeting.applyTargeting("flagA", input, {key: "bucketKeyA"})).toBe("red")
+  })
+
+  it("should evaluate valid rule with targeting key", () => {
+    const input = {
+      fractional: [
+        ["red", 50],
+        ["blue", 50]
+      ]
+    }
+
+    expect(targeting.applyTargeting("flagA", input, {targetingKey: "bucketKeyB"})).toBe("blue")
+  })
+
+  it("should validate rule - need at least two buckets", () => {
+    const input = {
+      fractional: [
+        ["red", 100],
+      ]
+    }
+
+    expect(targeting.applyTargeting("flagA", input, {targetingKey: "key"})).toBe(null)
+  })
+
+  it("should validate rule - buckets should add to 100", () => {
+    const input = {
+      fractional: [
+        ["red", 55],
+        ["blue", 55]
+      ]
+    }
+
+    expect(targeting.applyTargeting("flagA", input, {targetingKey: "key"})).toBe(null)
+  })
+
+  it("should validate rule - buckets should be variant and fraction", () => {
+    const input = {
+      fractional: [
+        ["red", 50],
+        [100, 50]
+      ]
+    }
```

Comment:
style: you could put all these in a "describe" func that is called "validation" and then just name them with what you've put after the `-`. This nesting is pretty idiomatic in JS tests.

Reply:
Good suggestion, will do this